### ISSUE TITLE
Generic type use , An error was obtained  --- Can't generate mapping method for a generic type variable source.

### ISSUE THREAD

**Issue by lixiao1995** [mapstruct/mapstruct#2954]
**background**

I want to mapping entity to VO  for PageInfo，The code :

```
@Mapper
public interface PageInfoConvert {
    PageInfoConvert INSTANCE = Mappers.getMapper(PageInfoConvert.class);

    @Mapping(target = "list",expression  = "java(voList)")
    <T,S> PageInfo<T> convert(PageInfo<S> pageInfo, @Context List<T> voList);
}
```

When compile, an error was obtained  ------ Can't generate mapping method for a generic type variable source.

**environment**
Jdk : 11
Mapstruct：1.5.0.RC1

I didn't get answer on the Internet
Please help me. Thanks

**Comment by Zegveld** [comment]
The problem is that mapstruct will attempt to map the `S` onto the `T` information, but it cannot determine what `T` or `S` is.

Perhaps the checking for generics can become less strict, when the fields affected by the generics are already explicitly specified. In your case it is most likely that inside `PageInfo<T>` exists a `List<T>` object.

**Comment by lixiao1995** [comment]
> 



> The problem is that mapstruct will attempt to map the `S` onto the `T` information, but it cannot determine what `T` or `S` is.
> 
> Perhaps the checking for generics can become less strict, when the fields affected by the generics are already explicitly specified. In your case it is most likely that inside `PageInfo<T>` exists a `List<T>` object.

**Comment by lixiao1995** [comment]
@Zegveld  You are right, The `List<T>`  exists. So, Can mapstruct support my demand?

**Comment by Zegveld** [comment]
It is currently not supported, but this is a nice feature to be added.

**Comment by lixiao1995** [comment]
> It is currently not supported, but this is a nice enhancement to be added.



> It is currently not supported, but this is a nice enhancement to be added.

**Comment by lixiao1995** [comment]
> It is currently not supported, but this is a nice enhancement to be added.

I'll look forward to it.thinks

**Comment by Zegveld** [comment]
please do not close this issue, as it will be used to track this.

**Comment by lixiao1995** [comment]
> please do not close this issue, as it will be used to track this.

got

**Comment by xg578962913** [comment]
hi, @Zegveld 
MapStruct seems to have to specify the specific class type of the JavaBean. Will this enhancement change this feature
Or this enhancement is only for JavaBean member variables, so that member variables can use generics, but JavaBean class  still need to specific class, and generics cannot be used
Like this?Such conversion will become possible?
`class JavaBean01<A> {
    A object1;
   ...
}
class JavaBean02<A>{
    A object1;
   ...
}`

Such conversion is still impossible?
<S, T> T mapper(S s, Class<T> t)

**Comment by filiphr** [comment]
@xg578962913 a generic conversion such as `<S, T> T mapper(S s, Class<T> t)` is not possible in MapStruct and we are not planning to implement something like that. 

If you are interested in something like that then you could implement it in your own way by implementing a custom processor that is going to create an implementation that works for something like this.

**Comment by xg578962913** [comment]
@filiphr Thanks, I got it

**Comment by filiphr** [comment]
@thedrow this looks like Kotlin code. How does the java code for this example look like?

**Comment by OYaagoub** [comment]
> @xg578962913 a generic conversion such as `<S, T> T mapper(S s, Class<T> t)` is not possible in MapStruct and we are not planning to implement something like that.
> 
> If you are interested in something like that then you could implement it in your own way by implementing a custom processor that is going to create an implementation that works for something like this.

stop closing your mind bro just in your ideas , it muy can me so powerfull bro in future is a good idea , you should take it in your side , you don't know what will happened hahahahha
